# EDA 
### Loading data set and exploring data content

In [3]:
import pandas as pd
import os

## Loading data into tables by table type

In [69]:
# load in each table type to create time-series; date from filename
# loading into each table type to simulate DWH
path = './src_data/'
filelist = os.listdir(path) 
print(filelist)
contracts_list = []
products_list = []
prices_list = []
for file in filelist:
    df2 = pd.read_csv(path+file, delimiter=';')
    date_from_name = file[0:8]
    # df2['Date'] = pd.to_datetime(date_from_name)
    if 'contracts' in file:
        contracts_list.append(df2)
    if 'products' in file:
        products_list.append(df2)
    if 'prices' in file:
        prices_list.append(df2)
contracts_df = pd.concat(contracts_list)
products_df = pd.concat(products_list)
prices_df = pd.concat(prices_list)
# for consistency across tables
products_df = products_df.rename(columns={'id':'productid'})
display(contracts_df.sort_values('startdate'))
display(products_df.sort_values('productid'))
display(prices_df.sort_values(['productid','pricecomponentid', 'valid_from']))

['20201101220039_prices.csv', '20201101230133_contracts.csv', '20201101220035_products.csv', '20201201220101_prices.csv', '20201201220059_contracts.csv', '20210101220045_products.csv', '20201201220034_products.csv', '20201001220144_contracts.csv', '20210101220104_contracts.csv', '20210101220057_prices.csv', '20201001220039_prices.csv', '20201001220037_products.csv']


id    type       energy   usage  usagenet   createdat   startdate  \
16199  16200  energy  electricity  2700.0    2860.0  2018-01-01  2018-01-01   
24981  24982  energy  electricity  2500.0    2442.0  2018-01-01  2018-01-01   
19890  19891  energy  electricity  2700.0       NaN  2018-01-01  2018-01-01   
24981  24982  energy  electricity  2500.0    2442.0  2018-01-01  2018-01-01   
11887  11888  energy  electricity  2700.0    2657.0  2018-01-01  2018-01-01   
...      ...     ...          ...     ...       ...         ...         ...   
33883  33884  energy  electricity  2100.0    2030.0  2020-12-20  2022-09-29   
31088  31089  energy  electricity  2300.0    2213.0  2020-10-16  2022-11-27   
31088  31089  energy  electricity  2300.0    2213.0  2020-10-16  2022-11-27   
30891  30892  energy  electricity  2000.0    2132.0  2020-10-20  2022-11-27   
30891  30892  energy  electricity  2000.0    2132.0  2020-10-20  2022-11-27   

          enddate fillingdatecancellation             cancellationreason  \
16199  2018-06-30              2018-06-02       fristgerecht durch Kunde   
24981  2020-03-31              2020-03-03       fristgerecht durch Kunde   
19890         NaN                     NaN                            NaN   
24981  2020-03-31              2020-03-03       fristgerecht durch Kunde   
11887  2019-05-31              2019-04-29  fristgerecht durch LichtBlick   
...           ...                     ...                            ...   
33883         NaN                     NaN                            NaN   
31088         NaN                     NaN                            NaN   
31088         NaN                     NaN                            NaN   
30891         NaN                     NaN                            NaN   
30891         NaN                     NaN                            NaN   

              city      status  productid modificationdate  
16199      Hamburg  terminated       1000       2018-06-02  
24981      Hamburg  terminated       1002       2019-08-30  
19890       Berlin  indelivery       1001       2019-08-13  
24981      Hamburg  terminated       1002       2019-08-30  
11887  Brandenburg  terminated       1002       2018-07-07  
...            ...         ...        ...              ...  
33883      Cologne      active       1000       2020-12-01  
31088    Stuttgart      active       1000       2020-11-05  
31088    Stuttgart      active       1000       2020-11-05  
30891     Mannheim      active       1000       2020-11-05  
30891     Mannheim      active       1000       2020-11-05  

[131600 rows x 14 columns]

productid productcode productname       energy        consumptiontype  \
0       1000      energy  ÖkoStrom A  electricity  household_electricity   
0       1000      energy  ÖkoStrom 1  electricity  household_electricity   
0       1000      energy  ÖkoStrom A  electricity  household_electricity   
0       1000      energy  ÖkoStrom A  electricity  household_electricity   
1       1001      energy  ÖkoStrom B  electricity  household_electricity   
1       1001      energy  ÖkoStrom B  electricity  household_electricity   
1       1001      energy  ÖkoStrom 2  electricity  household_electricity   
1       1001      energy  ÖkoStrom B  electricity  household_electricity   
2       1002      energy  ÖkoStrom C  electricity  household_electricity   
2       1002      energy  ÖkoStrom C  electricity  household_electricity   
2       1002      energy  ÖkoStrom 3  electricity  household_electricity   
2       1002      energy  ÖkoStrom C  electricity  household_electricity   
3       2001      energy    ÖkoGas A          gas              universal   
3       2001      energy    ÖkoGas 1          gas              universal   
3       2001      energy    ÖkoGas A          gas              universal   
3       2001      energy    ÖkoGas A          gas              universal   
4       2002      energy    ÖkoGas B          gas              universal   
4       2002      energy    ÖkoGas 2          gas              universal   
4       2002      energy    ÖkoGas B          gas              universal   
4       2002      energy    ÖkoGas B          gas              universal   
5       9001      energy   ÖkoTest A  electricity  household_electricity   
5       9001      energy   ÖkoTest A  electricity  household_electricity   
5       9001      energy   ÖkoTest A  electricity  household_electricity   

   deleted modificationdate  
0        0       2020-08-27  
0        0       2020-12-29  
0        0       2020-08-27  
0        0       2020-08-27  
1        0       2020-02-05  
1        0       2020-02-05  
1        0       2020-12-29  
1        0       2020-02-05  
2        0       2020-01-01  
2        0       2020-06-07  
2        0       2020-12-29  
2        0       2020-01-01  
3        0       2019-05-21  
3        0       2020-12-29  
3        0       2019-05-21  
3        0       2019-05-21  
4        0       2020-06-04  
4        0       2020-12-29  
4        0       2020-06-04  
4        0       2020-06-04  
5        1       2020-12-02  
5        0       2020-11-29  
5        1       2020-12-02

id  productid  pricecomponentid productcomponent  price    unit  \
0    1       1000                 1        baseprice   8.99  €/year   
0    1       1000                 1        baseprice   8.99  €/year   
0    1       1000                 1        baseprice   8.99  €/year   
0    1       1000                 1        baseprice   8.99  €/year   
12  13       1000                 1        baseprice   9.79  €/year   
12  13       1000                 1        baseprice   9.79  €/year   
12  13       1000                 1        baseprice   9.79  €/year   
12  13       1000                 1        baseprice   9.79  €/year   
1    2       1000                 2     workingprice  29.80  ct/kwh   
1    2       1000                 2     workingprice  29.80  ct/kwh   
1    2       1000                 2     workingprice  29.80  ct/kwh   
1    2       1000                 2     workingprice  29.80  ct/kwh   
13  14       1000                 2     workingprice  29.90  ct/kwh   
13  14       1000                 2     workingprice  29.90  ct/kwh   
13  14       1000                 2     workingprice  29.90  ct/kwh   
13  14       1000                 2     workingprice  29.90  ct/kwh   
2    3       1001                 1        baseprice   9.99  €/year   
2    3       1001                 1        baseprice   9.99  €/year   
2    3       1001                 1        baseprice   9.99  €/year   
2    3       1001                 1        baseprice   9.99  €/year   
4    5       1001                 2     workingprice  29.40  ct/kwh   
4    5       1001                 2     workingprice  29.40  ct/kwh   
4    5       1001                 2     workingprice  29.40  ct/kwh   
4    5       1001                 2     workingprice  29.40  ct/kwh   
11  12       1001                 2     workingprice  29.30  ct/kwh   
11  12       1001                 2     workingprice  29.30  ct/kwh   
11  12       1001                 2     workingprice  29.30  ct/kwh   
11  12       1001                 2     workingprice  29.30  ct/kwh   
3    4       1002                 1        baseprice   9.49  €/year   
3    4       1002                 1        baseprice   9.49  €/year   
3    4       1002                 1        baseprice   9.49  €/year   
3    4       1002                 1        baseprice   9.49  €/year   
5    6       1002                 2     workingprice  28.90  ct/kwh   
5    6       1002                 2     workingprice  28.90  ct/kwh   
5    6       1002                 2     workingprice  28.90  ct/kwh   
5    6       1002                 2     workingprice  28.90  ct/kwh   
10  11       1002                 2     workingprice  29.10  €/year   
10  11       1002                 2     workingprice  29.10  €/year   
10  11       1002                 2     workingprice  29.10  €/year   
10  11       1002                 2     workingprice  29.10  €/year   
6    7       2001                 1        baseprice  13.99  €/year   
6    7       2001                 1        baseprice  13.99  €/year   
6    7       2001                 1        baseprice  13.99  €/year   
6    7       2001                 1        baseprice  13.99  €/year   
14  15       2001                 1        baseprice  14.29  €/year   
7    8       2001                 2     workingprice   4.90  ct/kwh   
7    8       2001                 2     workingprice   4.90  ct/kwh   
7    8       2001                 2     workingprice   4.90  ct/kwh   
7    8       2001                 2     workingprice   4.90  ct/kwh   
15  16       2001                 2     workingprice   5.15  ct/kwh   
8    9       2002                 1        baseprice  13.99  €/year   
8    9       2002                 1        baseprice  13.99  €/year   
8    9       2002                 1        baseprice  13.99  €/year   
8    9       2002                 1        baseprice  13.99  €/year   
16  17       2002                 1        baseprice  14.19  €/year   
9   10       2002           

### Notes, assumptions, questions:
- enddate of NaN means the contract has not been terminated
- what is date in filename? when data was provided, but does not seem related to any other date
- there are apparently duplicate rows, need to remove
- check for missing data
- look for outliers in useage, dates, ...
- isolate most recent product names
- 

## Checking data and cleaning

#### Getting the current prices and products

In [67]:
prices_current_df = prices_df.sort_values(by=['productid',
                                              'pricecomponentid', 
                                              'valid_from', 
                                              'modificationdate']).drop_duplicates(subset=['productid',
                                                                                           'pricecomponentid', 
                                                                                           'valid_from'], 
                                                                                           keep="last")
prices_current_df.sort_values(by=['productid',
                                              'pricecomponentid', 
                                              'valid_from', 
                                              'modificationdate'])

id  productid  pricecomponentid productcomponent  price    unit  \
0    1       1000                 1        baseprice   8.99  €/year   
12  13       1000                 1        baseprice   9.79  €/year   
1    2       1000                 2     workingprice  29.80  ct/kwh   
13  14       1000                 2     workingprice  29.90  ct/kwh   
2    3       1001                 1        baseprice   9.99  €/year   
4    5       1001                 2     workingprice  29.40  ct/kwh   
11  12       1001                 2     workingprice  29.30  ct/kwh   
3    4       1002                 1        baseprice   9.49  €/year   
5    6       1002                 2     workingprice  28.90  ct/kwh   
10  11       1002                 2     workingprice  29.10  €/year   
6    7       2001                 1        baseprice  13.99  €/year   
14  15       2001                 1        baseprice  14.29  €/year   
7    8       2001                 2     workingprice   4.90  ct/kwh   
15  16       2001                 2     workingprice   5.15  ct/kwh   
8    9       2002                 1        baseprice  13.99  €/year   
16  17       2002                 1        baseprice  14.19  €/year   
9   10       2002                 2     workingprice   5.20  ct/kwh   
17  18       2002                 2     workingprice   5.25  ct/kwh   

    valid_from valid_until modificationdate  
0   2019-01-01  2020-05-31       2020-05-24  
12  2020-06-01  9999-12-31       2020-05-24  
1   2019-01-01  2020-05-31       2020-05-24  
13  2020-06-01  9999-12-31       2020-05-24  
2   2019-01-01  9999-12-31       2018-12-23  
4   2019-01-01  2019-12-31       2019-12-22  
11  2020-01-01  9999-12-31       2019-12-22  
3   2019-01-01  9999-12-31       2018-12-23  
5   2019-01-01  2019-12-31       2019-12-22  
10  2020-01-01  9999-12-31       2019-12-22  
6   2019-01-01  2020-12-31       2020-12-28  
14  2021-01-01  9999-12-31       2020-12-28  
7   2019-01-01  2020-12-31       2020-12-28  
15  2021-01-01  9999-12-31       2020-12-28  
8   2019-01-01  2020-12-31       2020-12-28  
16  2021-01-01  9999-12-31       2020-12-28  
9   2019-01-01  2020-12-31       2020-12-28  
17  2021-10-01  9999-12-31       2020-12-28

In [68]:
products_current_df = products_df.sort_values(by=['productid',
                                              'energy',  
                                              'modificationdate']).drop_duplicates(subset=['productid',
                                                                                           'energy'],
                                                                                           keep="last")
products_current_df.sort_values(by=['productid',
                                              'energy',  
                                              'modificationdate'])

productid productcode productname       energy        consumptiontype  \
0       1000      energy  ÖkoStrom 1  electricity  household_electricity   
1       1001      energy  ÖkoStrom 2  electricity  household_electricity   
2       1002      energy  ÖkoStrom 3  electricity  household_electricity   
3       2001      energy    ÖkoGas 1          gas              universal   
4       2002      energy    ÖkoGas 2          gas              universal   
5       9001      energy   ÖkoTest A  electricity  household_electricity   

   deleted modificationdate  
0        0       2020-12-29  
1        0       2020-12-29  
2        0       2020-12-29  
3        0       2020-12-29  
4        0       2020-12-29  
5        1       2020-12-02

#### Check if product_ids are in all tables

In [71]:
products_df.productid.unique(), prices_df.productid.unique(), contracts_df.productid.unique()

(array([1000, 1001, 1002, 2001, 2002, 9001]),
 array([1000, 1001, 1002, 2001, 2002]),
 array([1000, 1002, 1001, 2000, 2001]))

- 9001 has no price info and no contracts connected to it

#### Remove duplicate contracts

In [72]:
print(contracts_df.shape)
contracts_df = contracts_df.drop_duplicates()
print(contracts_df.shape)


(131600, 14)
(37393, 14)


In [59]:
prices_df.sort_values(['productid','pricecomponentid', 'valid_from'])

id  productid  pricecomponentid productcomponent  price    unit  \
0    1       1000                 1        baseprice   8.99  €/year   
12  13       1000                 1        baseprice   9.79  €/year   
1    2       1000                 2     workingprice  29.80  ct/kwh   
13  14       1000                 2     workingprice  29.90  ct/kwh   
2    3       1001                 1        baseprice   9.99  €/year   
4    5       1001                 2     workingprice  29.40  ct/kwh   
11  12       1001                 2     workingprice  29.30  ct/kwh   
3    4       1002                 1        baseprice   9.49  €/year   
5    6       1002                 2     workingprice  28.90  ct/kwh   
10  11       1002                 2     workingprice  29.10  €/year   
6    7       2001                 1        baseprice  13.99  €/year   
6    7       2001                 1        baseprice  13.99  €/year   
14  15       2001                 1        baseprice  14.29  €/year   
7    8       2001                 2     workingprice   4.90  ct/kwh   
7    8       2001                 2     workingprice   4.90  ct/kwh   
15  16       2001                 2     workingprice   5.15  ct/kwh   
8    9       2002                 1        baseprice  13.99  €/year   
8    9       2002                 1        baseprice  13.99  €/year   
16  17       2002                 1        baseprice  14.19  €/year   
9   10       2002                 2     workingprice   5.20  ct/kwh   
9   10       2002                 2     workingprice   5.20  ct/kwh   
17  18       2002                 2     workingprice   5.25  ct/kwh   

    valid_from valid_until modificationdate  
0   2019-01-01  2020-05-31       2020-05-24  
12  2020-06-01  9999-12-31       2020-05-24  
1   2019-01-01  2020-05-31       2020-05-24  
13  2020-06-01  9999-12-31       2020-05-24  
2   2019-01-01  9999-12-31       2018-12-23  
4   2019-01-01  2019-12-31       2019-12-22  
11  2020-01-01  9999-12-31       2019-12-22  
3   2019-01-01  9999-12-31       2018-12-23  
5   2019-01-01  2019-12-31       2019-12-22  
10  2020-01-01  9999-12-31       2019-12-22  
6   2019-01-01  9999-12-31       2018-12-23  
6   2019-01-01  2020-12-31       2020-12-28  
14  2021-01-01  9999-12-31       2020-12-28  
7   2019-01-01  9999-12-31       2018-12-23  
7   2019-01-01  2020-12-31       2020-12-28  
15  2021-01-01  9999-12-31       2020-12-28  
8   2019-01-01  9999-12-31       2018-12-23  
8   2019-01-01  2020-12-31       2020-12-28  
16  2021-01-01  9999-12-31       2020-12-28  
9   2019-01-01  9999-12-31       2018-12-23  
9   2019-01-01  2020-12-31       2020-12-28  
17  2021-10-01  9999-12-31       2020-12-28

#### Look for missing data fields that are necessary

In [36]:
contracts_df[contracts_df.type.isna() | contracts_df.energy.isna() ]



Empty DataFrame
Columns: [id, type, energy, usage, usagenet, createdat, startdate, enddate, fillingdatecancellation, cancellationreason, city, status, productid, modificationdate]
Index: []

In [40]:
contracts_df[contracts_df.usage.isna() ]

Empty DataFrame
Columns: [id, type, energy, usage, usagenet, createdat, startdate, enddate, fillingdatecancellation, cancellationreason, city, status, productid, modificationdate]
Index: []

In [39]:
contracts_df[contracts_df.usagenet.isna() ]


id    type       energy    usage  usagenet   createdat   startdate  \
20        21  energy  electricity   2400.0       NaN  2019-06-17  2019-10-01   
29        30  energy  electricity   2500.0       NaN  2020-09-12  2020-10-01   
34        35  energy  electricity   1900.0       NaN  2018-07-03  2019-03-01   
41        42  energy  electricity   2600.0       NaN  2020-01-06  2020-03-01   
49        50  energy  electricity   2600.0       NaN  2019-10-12  2020-03-01   
...      ...     ...          ...      ...       ...         ...         ...   
19335  19336  energy  electricity   2100.0       NaN  2020-08-14  2022-07-29   
20647  20648  energy          gas  11400.0       NaN  2020-07-23  2021-06-01   
21416  21417  energy          gas  11700.0       NaN  2020-09-22  2021-07-29   
22880  22881  energy  electricity   2300.0       NaN  2020-05-11  2022-06-29   
27236  27237  energy          gas   8700.0       NaN  2020-09-25  2021-07-30   

      enddate fillingdatecancellation cancellationreason          city  \
20        NaN                     NaN                NaN       Hamburg   
29        NaN                     NaN                NaN        Berlin   
34        NaN                     NaN                NaN        Berlin   
41        NaN                     NaN                NaN      Nordhorn   
49        NaN                     NaN                NaN         Herne   
...       ...                     ...                ...           ...   
19335     NaN                     NaN                NaN      Duisburg   
20647     NaN                     NaN                NaN     Wiesbaden   
21416     NaN                     NaN                NaN  Braunschweig   
22880     NaN                     NaN                NaN       Cologne   
27236     NaN                     NaN                NaN        Lübeck   

           status  productid modificationdate  
20     indelivery       1001       2020-05-01  
29     indelivery       1000       2020-09-26  
34     indelivery       1002       2020-01-03  
41     indelivery       1000       2020-06-11  
49     indelivery       1001       2020-01-09  
...           ...        ...              ...  
19335      active       1002       2020-12-01  
20647      active       2000       2020-12-01  
21416      active       2000       2020-12-01  
22880      active       1002       2020-12-01  
27236      active       2000       2020-12-01  

[2453 rows x 14 columns]

In [37]:
contracts_df[contracts_df.startdate.isna()]


Empty DataFrame
Columns: [id, type, energy, usage, usagenet, createdat, startdate, enddate, fillingdatecancellation, cancellationreason, city, status, productid, modificationdate]
Index: []

In [38]:
contracts_df[ contracts_df.productid.isna()]

Empty DataFrame
Columns: [id, type, energy, usage, usagenet, createdat, startdate, enddate, fillingdatecancellation, cancellationreason, city, status, productid, modificationdate]
Index: []

- "usagenet" missing for many accounts; for simplicity just use "usage" and not "usagenet